Update formulae to use Utils.safe_popen_read

Rylan12 opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

See https://github.com/Homebrew/brew/pull/7695.

This PR updates formulae to follow to the new RuboCop check for using `Utils.safe_popen_read` instead of `Utils.popen_read`

I put everything in one commit because it was easy to just run `brew style --fix` on the Formula directory.

There are no revision bumps right now but I can add those if desired

Rylan12 commented:
Tagging @MikeMcQuaid

Rylan12 commented:
The 10.15 tests haven't completed yet, but it looks like these are the formulae with issues (I've specified if the issue is not an install issue):
- [ ] checkstyle (tests)
- [ ] gitlab-runner
- [ ] goofys
- [ ] helm
- [ ] helm@2
- [ ] inlets
- [ ] jabba
- [ ] kamel
- [ ] kapacitor
- [ ] krew
- [ ] kustomize
- [ ] linkerd
- [ ] minikube
- [ ] mmctl
- [ ] mockserver (tests)
- [ ] natalie (tests)
- [ ] neovim (audit)
- [ ] oauth2_proxy (tests)
- [ ] skopeo
- [ ] sync_gateway
- [ ] wiremock-standalone (tests)

If there's something I can do to speed up the tests so it doesn't take another 10 hours let me know.

Rylan12 commented:
Also, I welcome thoughts/ideas for fixing the ~20 that don't work. I'll start to go through them when I'm available but it will probably take a while.

Rylan12 commented:
Actually, looking at the logs, many of the formulae fail with this message:

```
Error: Failed to download resource "gitlab-runner"
Failure while executing; `git clone --branch v13.0.1 https://gitlab.com/gitlab-org/gitlab-runner.git /Users/brew/Library/Caches/Homebrew/gitlab-runner--git` exited with . Here's the output:
dyld: Library not loaded: /usr/local/opt/pcre2/lib/libpcre2-8.0.dylib
  Referenced from: /usr/local/bin/git
  Reason: image not found
```

`gitlab-runner` (just the first example) builds correctly on my machine, so this seems like it might be a CI error of some sort.

MikeMcQuaid commented on Formula/armadillo.rb:
```diff
@@ -34,6 +34,6 @@ def install
     EOS
     system ENV.cxx, "test.cpp", "-I#{include}", "-L#{lib}", "-larmadillo", "-o", "test"
-    assert_equal Utils.popen_read("./test").to_i, version.to_s.to_i
+    assert_equal Utils.safe_popen_read("./test").to_i, version.to_s.to_i
```
Any cases in the `test do` block should be `shell_output` instead (which will produce nicer error messages for tests). This could perhaps be an extension of the rubocop?

MikeMcQuaid commented on Formula/botan.rb:
```diff
@@ -38,7 +38,7 @@ def install
   test do
     (testpath/"test.txt").write "Homebrew"
-    (testpath/"testout.txt").write Utils.popen_read("#{bin}/botan base64_enc test.txt")
+    (testpath/"testout.txt").write Utils.safe_popen_read("#{bin}/botan base64_enc test.txt")
```
Should also be `shell_output`

MikeMcQuaid commented on Formula/capnp.rb:
```diff
@@ -25,7 +25,7 @@ def install
     text = "\"Is a happy little duck\""
 
-    file.write Utils.popen_read("#{bin}/capnp id").chomp + ";\n"
+    file.write Utils.safe_popen_read("#{bin}/capnp id").chomp + ";\n"
```
Should also be `shell_output`

MikeMcQuaid commented on Formula/checkstyle.rb:
```diff
@@ -15,7 +15,7 @@ def install
     path.write "public class Foo{ }\n"
 
-    output = Utils.popen_read("#{bin}/checkstyle -c /sun_checks.xml #{path}")
+    output = Utils.safe_popen_read("#{bin}/checkstyle -c /sun_checks.xml #{path}")
```
Should also be `shell_output`

MikeMcQuaid commented on Formula/elektra.rb:
```diff
@@ -33,7 +33,7 @@ def install
     output = shell_output("#{bin}/kdb get system/elektra/version/infos/licence")
     assert_match "BSD", output
-    Utils.popen_read("#{bin}/kdb", "plugin-list").split.each do |plugin|
+    Utils.safe_popen_read("#{bin}/kdb", "plugin-list").split.each do |plugin|
```
Should also be `shell_output`

MikeMcQuaid commented on Formula/jabba.rb:
```diff
@@ -32,7 +32,7 @@ def install
     ENV["JABBA_HOME"] = testpath/"jabba_home"
     system bin/"jabba", "install", "1.13.0"
-    jdk_path = Utils.popen_read("#{bin}/jabba which 1.13.0").strip
+    jdk_path = Utils.safe_popen_read("#{bin}/jabba which 1.13.0").strip
```
Should also be `shell_output`

MikeMcQuaid commented:
Thanks for the PR @Rylan12!

SMillerDev commented:
This PR runs for over a day to test everything, please make sure all issues are resolved locally before pushing again. And make sure to cancel the old run before you push again.

Rylan12 commented:
Ahh, oops. Just pushed again. I was trying to avoid having to install all the formulae on my machine but maybe I'll have to. How do I cancel?

SMillerDev commented:
You should be able to see a `cancel workflow` button on the top right corner when you got to `checks`. If it isn't there you should ask a maintainer.

I know the problem of not wanting to clog up your machine but this is just a lot formula that need building (also as dependencies)

Rylan12 commented:
I'm not seeing it, so if you could cancel I'd appreciate it.

Rylan12 commented:
I'll do what I can on my machine but (as I mentioned in another [comment](https://github.com/Homebrew/homebrew-core/pull/55934#issuecomment-640994804)), some of the issues that I found were not reproducible on my local machine, so I'm not sure what to do about those.

Rylan12 commented:
I looked into it a little more. `pcre2` is uninstalled after the `git` formula is tested (see [line 624 of the 10.14 log here](https://github.com/Homebrew/homebrew-core/pull/55934/checks?check_run_id=750516456#step:6:624) or below) so all formulae tested after `git` don't have `pcre2` installed.

<details>
<summary><code>git</code> formula test log</summary>
<pre><code>==> Running Formulae#formula!(git)
==> brew fetch --retry pcre2
==> brew fetch --retry git --build-bottle --force
==> brew install --only-dependencies --verbose --build-bottle git
==> brew install --verbose --build-bottle git
==> brew audit --skip-style git --online
==> brew bottle --verbose --json git
==> brew bottle --merge --write --no-commit ./git--2.27.0.mojave.bottle.json
==> brew uninstall --force git
==> brew install --only-dependencies ./git--2.27.0.mojave.bottle.1.tar.gz
==> brew install ./git--2.27.0.mojave.bottle.1.tar.gz
==> brew linkage --test git
==> brew install --only-dependencies --include-test git
==> brew test --verbose git
<strong>==> brew uninstall --force pcre2</strong ></code></pre>
</details>

If I have `git` in its own commit will its tests run separately? That way, `pcre2` wouldn't be uninstalled, affecting the other formulae. If not, I could make a separate PR for `git`.

MikeMcQuaid commented:
> * **neovim**: needs to be updated so it doesn't reference the homebrew shims directory. _Not sure exactly how to proceed._

@bo98 may be able to provide guidance here. This isn't a block on the PR being merged, to be clear.



> If not, I could make a separate PR for `git`.'

This seems like the best idea. Yeh, you need them to be in separate PRs not just separate commits.

Rylan12 commented:
> This isn't a block on the PR being merged, to be clear.

Just clarifying: this means that I don't have to fix the neovim audit issue in this PR for the checks to pass?

In the CI logs, it lists `brew audit --skip-style neovim --online` as a failure so I just want to make sure that if I don't change it the CI checks will still pass (and it won't need to run for an additional 10 hours)

Bo98 commented:
`strings <bin>/nvim | grep shims` can be a start for debugging - I'll have a look properly tomorrow.

Uninstalling pcre2 is intentional behaviour. The issue is that `brew` starts brewed git for fetching when it was using system git prior to brewed git being installed, and test-bot isn't expecting that (test-bot _intentionally_ uninstalls dependencies as it doesn't think the formula will be used again). This used to be an issue even when testing pull requests with git on its own, and we fixed this in homebrew-test-bot by enforcing use of system git. We could probably make that apply universally across brew by setting `HOMEBREW_GIT_PATH` somewhere in homebrew-test-bot.

MikeMcQuaid commented:
> Just clarifying: this means that I don't have to fix the neovim audit issue in this PR for the checks to pass?

It means we don't need to have passing CI in this case.



> We could probably make that apply universally across brew by setting `HOMEBREW_GIT_PATH` somewhere in homebrew-test-bot.

I thought we did do that? It might need extended, though!

Bo98 commented:
> I thought we did do that? It might need extended, though!

What we did is swap out and `git` invokations in homebrew-test-bot itself with system git. This was sufficient to make a git pull request pass.

The changes didn't however cover anything that `brew` itself does - which can happen if git isn't in a pull request on its own (e.g. fetching a git repo of another formula tested after git).

MikeMcQuaid commented:
> The changes didn't however cover anything that `brew` itself does - which can happen if git isn't in a pull request on its own (e.g. fetching a git repo of another formula tested after git).

My understand is that Homebrew should always use the system Git unless we set `HOMEBREW_GIT`. If/when this isn't the case: yeh, this is a bug and we should fix it 👍

Bo98 commented:
I think it uses whatever is in the PATH by default, which brewed git will be.

Rylan12 commented:
I see that the crystal formula has been updated, so I'll rebase and make sure to fix any conflict there.

Aside from that, here are the other formulae that were failing:
- **jabba + skopeo**: moved to #56057
- **natalie**: moved to #56162
- **sync_gateway**: Unable to be built currently. See #55219. It looks like a fix is in progress upstream in https://github.com/couchbase/sync_gateway/pull/4610. I guess at this point I'll just wait and hope they fix that upstream relatively quickly. To allow this PR to move forward, I will move the natalie changes to its own PR

Everything else (except neovim) has been fixed locally or was related to the git/pcre2 issue

Rylan12 commented:
**boost-python** is also giving me some trouble.

`brew test boost-python` output:
```
Testing boost-python
==> python-config --includes
==> python-config --ldflags
==> /usr/bin/clang++ -shared hello.cpp -L/usr/local/Cellar/boost-python/1.72.0/lib -lboost_python27 -o hello.so -I/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/include/python2.7 -I/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/include/python3.7m -I/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/include/python3.7m -L/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/lib/python3.7/config-3.7m-darwin -lpython3.7m -ldl -framework CoreFoundation
Last 15 lines from /Users/rylanpolster/Library/Logs/Homebrew/boost-python/test.01.clang++:
hello.so
-I/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/include/python2.7
-I/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/include/python3.7m
-I/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/include/python3.7m
-L/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/lib/python3.7/config-3.7m-darwin
-lpython3.7m
-ldl
-framework
CoreFoundation

Undefined symbols for architecture x86_64:
  "boost::python::detail::init_module(PyModuleDef&, void (*)())", referenced from:
      _PyInit_hello in hello-f24fe4.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
An exception occurred within a child process:
  BuildError: Failed executing: /usr/bin/clang++ -shared hello.cpp -L/usr/local/Cellar/boost-python/1.72.0/lib -lboost_python27 -o hello.so -I/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/include/python2.7 -I/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/include/python3.7m -I/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/include/python3.7m -L/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/lib/python3.7/config-3.7m-darwin -lpython3.7m -ldl -framework CoreFoundation
/usr/local/Homebrew/Library/Homebrew/formula.rb:2013:in `block in system'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1951:in `open'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1951:in `system'
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/boost-python.rb:84:in `block in <class:BoostPython>'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1821:in `block (3 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/utils.rb:474:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1820:in `block (2 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/formula.rb:888:in `with_logging'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1819:in `block in run_test'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2063:in `block in mktemp'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:57:in `block in run'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:57:in `chdir'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:57:in `run'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2062:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1813:in `run_test'
/usr/local/Homebrew/Library/Homebrew/test.rb:35:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
/usr/local/Homebrew/Library/Homebrew/test.rb:34:in `<main>'
```

Note that the formula builds correctly. This issue is not related to the changes made in this PR, so I will plan on removing changes to boost-python from this PR. Additionally, https://github.com/Homebrew/brew/pull/7695 won't check for `Utils.popen_read` in the `test` block, so changing the `Utils.popen_read` to `shell_output` in boost-python would not be a necessary change (but might still be a good idea).

Rylan12 commented:
I have fixed everything on my local machine, so I have pushed the changes so the CI check can run overnight (for me).

I've commented about these a few times before, but I want here is a list of formulae that I want to update but, due to other issues, decided not to include in this PR:
- **boost-python**: I wasn't able to build this locally. The `Utils.popen_read` is in the `test` block so, while it may still be good to change it, I didn't think it was worth the hassle to debug and fix for this PR because it isn't blocking https://github.com/Homebrew/brew/pull/7695.
- **sync-gateway**: I wasn't able to build this locally. they are working on porting to Python 3 upstream, so I'm hoping that will be completed soon (and fix some of these issues). This formula does block https://github.com/Homebrew/brew/pull/7695, but I think that a change could be committed directly without any negative consequences (given that the formula doesn't currently build)

Many thanks to @MikeMcQuaid, @Bo98, and @SMillerDev for their help on this monster PR.

SMillerDev commented:
Thanks for all the work on your side too, it's much appreciated.

Bo98 commented:
> `-I/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/include/python2.7`

Yeah, `boost-python` seems to be in a bit of a mess there. To be clear: it should be using Python 2 only.

> sync-gateway: I wasn't able to build this locally.

Yep, known failure.

> I think that a change could be committed directly without any negative consequences (given that the formula doesn't currently build)

Fine with me.

Rylan12 commented:
It looks like there were some CI errors that weren't expected (i.e. not neovim). I'll take a look later today to see if I can find the problem(s). At first glance, it looks like the formulae that had issues this time were different than the formulae that had issues last time; I'm not sure why that would happen exactly.

Rylan12 commented:
Here is the latest update:

## Issues I know how to (and will) fix
- **glassfish**: used the deprecated `Language::Java.java_home_cmd`. I'll make a PR to fix this separately and remove **glassfish** from this PR
- **kamel**: It looks like the completion generation code doesn't work, so `Utils.safe_popen_read` is failing. Interestingly enough, this is the motivation for making this whole PR—to catch these issues. I'll open another PR to work on this and remove the changes from here.

## Issues I could use some help/advice on
- **boost-python3**, **php@7.2**, **php@7.3**: These fail the audit saying that `revision 1` should be removed. Maybe I'm not completely understanding what revision does but I would think that the `revision 1` should be kept so that users still on `revision 0` (i.e. no revision line) will upgrade. Users already on `revision 1` shouldn't need to upgrade though.
- **faas-cli**: The tests failed for 10.14. On my machine (10.15) all tests pass. The only change made in **faas-cli** was changing `Utils.popen_read("git", "rev-parse", "HEAD").chomp` to `Utils.safe_popen_read("git", "rev-parse", "HEAD").chomp`. I've checked and these commands both return the same string, so I don't know what the issue is.

Rylan12 commented:
@MikeMcQuaid or @Bo98 can you take another look at this? I'm not sure what to do about the remaining issues

SMillerDev commented:
LGTM, not sure what is triggering the php issues but it looks fine to me. For faas I guess we'll see what happens the next time it needs to run through CI.

MikeMcQuaid commented:
> * These fail the audit saying that `revision 1` should be removed.

This is maybe a bug with `brew audit`. You can ignore it.



> * I've checked and these commands both return the same string, so I don't know what the issue is.

This seems likely to be an existing failure so can be ignored for now.

I've reset these commits locally and rebuilt them and pushed. Note for future we need one commit per formula changed.

Thanks again @Rylan12, great work!

Rylan12 commented:
Thanks everyone!
